test/auth.test.ts

```diff
@@ -0,0 +1,356 @@
+    let onTokenFetchedSpy: any = jest.fn();
+
+    beforeEach(() => {
+      auth = new Auth({
+        clientId,
+        host,
+        messenger: new Messenger({
+          sharedWorkerUrl: '/worker.js',
+        }),
+        _onTokenFetched: onTokenFetchedSpy,
+      });
+      storageGetSpy = jest.spyOn(Storage.prototype, 'getItem');
+      storageSetSpy = jest.spyOn(Storage.prototype, 'setItem');
+      storageRemoveSpy = jest.spyOn(Storage.prototype, 'removeItem');
+    });
+
+    afterEach(() => {
+      auth = undefined as any;
+      storageGetSpy.mockRestore();
+      storageSetSpy.mockRestore();
+      storageRemoveSpy.mockRestore();
+    });
+
+    it.each([
+      ['whatever', true],
+      [null, false],
+    ])('calls getState to determine if redirected', (state, expected) => {
+      storageGetSpy.mockImplementation(() => state);
+      expect(auth.isRedirected()).toBe(expected);
+      expect(storageGetSpy).toBeCalled();
+    });
+
+    it('does nothing if already handled redirect', () => {
+      auth['didHandleRedirect'] = true;
+      auth.handleRedirect();
+      expect(storageGetSpy).not.toBeCalled();
+    });
+
+    it('rejects if state not equal', async () => {
+      delete (global.window as any).location;
+      global.window.location = {
+        search: '?state=somethingelse',
+      } as any;
+      storageGetSpy.mockImplementation(() => 'something');
+
+      await expect(auth.handleRedirect()).rejects.toThrowError();
+      expect(storageRemoveSpy).toHaveBeenCalled();
+    });
+
+    it('makes a request with auth code', async () => {
+      const state = 'state';
+      const authCode = '1234';
+      delete (global.window as any).location;
+      global.window.location = {
+        search: `?state=${state}&authorization_code=${authCode}`,
+      } as any;
+      storageGetSpy.mockImplementation(() => state);
+      const consoleSpy = jest.spyOn(console, 'error');
+      consoleSpy.mockImplementation(() => {});
+      global.fetch = jest.fn();
+
+      expect.assertions(1);
+      await auth.handleRedirect();
+
+      expect(global.fetch).toHaveBeenCalledWith(
+        `${host}/oauth/token`,
+        expect.objectContaining({
+          body: JSON.stringify({ authCode }),
+        })
+      );
+      consoleSpy.mockRestore();
+    });
+
+    describe('response handling', () => {
+      let consoleSpy: any;
+
+      beforeEach(() => {
+        const state = 'state';
+        delete (global.window as any).location;
+        global.window.location = {
+          search: `?state=${state}`,
+        } as any;
+        storageGetSpy.mockImplementation(() => state);
+        consoleSpy = jest.spyOn(console, 'error');
+        consoleSpy.mockImplementation(() => {});
+      });
+
+      afterEach(() => {
+        consoleSpy.mockRestore();
+      });
+
+      it('shows error, deletes state, and sends message on bad response', async () => {
+        const messengerSpy = jest.spyOn(auth['messenger'], 'sendWorkerMessage');
```

Comment:
```suggestion
        const messengerSpy = jest.spyOn(auth.messenger, 'sendWorkerMessage');
```

Reply:
ah yeah - it's to access private members without typescript complaining 🤷